src/java/test/org/apache/zookeeper/test/KeyAuthClientTest.java

```diff
@@ -0,0 +1,131 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.test;
+
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.ZooDefs.Ids;
+import org.apache.zookeeper.ZooKeeper;
+import org.apache.zookeeper.data.ACL;
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+public class KeyAuthClientTest extends ClientBase {
+    private static final Logger LOG = LoggerFactory.getLogger(KeyAuthClientTest.class);
+    static {
+        System.setProperty("zookeeper.authProvider.1","org.apache.zookeeper.server.auth.KeyAuthenticationProvider");
+    }
+
+    public void createNodePrintAcl(ZooKeeper zk, String path, String testName) {
+      try {
+        LOG.debug("KeyAuthenticationProvider Creating Test Node:"+path+".\n");
+        zk.create(path, null, Ids.CREATOR_ALL_ACL, CreateMode.PERSISTENT);
+        List<ACL> acls = zk.getACL(path, null);
+        LOG.debug("Node: "+path+" Test:"+testName+" ACLs:");
+        for (ACL acl : acls) {
+          LOG.debug("  "+acl.toString());
+        }
+      } catch (Exception e) {
+          LOG.debug("  EXCEPTION THROWN", e);
+      }
+    }
+
+    public void testPreAuth() throws Exception {
+        ZooKeeper zk = createClient();
+        zk.addAuthInfo("key", "25".getBytes());
+        try {
+            createNodePrintAcl(zk, "/pre", "testPreAuth");
+            zk.setACL("/", Ids.CREATOR_ALL_ACL, -1);
+            zk.getChildren("/", false);
+            zk.create("/abc", null, Ids.CREATOR_ALL_ACL, CreateMode.PERSISTENT);
+            zk.setData("/abc", "testData1".getBytes(), -1);
+            zk.create("/key", null, Ids.CREATOR_ALL_ACL, CreateMode.PERSISTENT);
+            zk.setData("/key", "5".getBytes(), -1);
+            Thread.sleep(1000);
```

Comment:
Only thing I can suppose (and **please** correct me if I am saying something **utterly stupid**) is that this ``sleep`` was inserted in the hopes of giving some time to ZK commit the changes. Naive but yet (shrug)... 

Well, I guess we can either remove it or replace it with a ``zk.getData("/key");``, right?